db/badger_db.go

```diff
@@ -0,0 +1,323 @@
+// Copyright 2017 Tendermint. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// BadgerDB enables us the use of BadgerDB a fast key value store in pure Go.
+// Typical usage follows:
+//  db, err := db.NewBadgerDB("badger", targetDir)
+//  if err != nil {
+//	log.Fatalf("badgerDB initialization: %v", err)
+//  }
+//  defer db.Close()
+//
+//  For the asynchronous set
+//  db.Set([]byte("foo"), []byte("bar"))
+//  db.SetSync([]byte("true"), []byte("tendermint"))
+//  db.Delete([]byte("true"))
+//  db.Deletesync([]byte("true"))
+//  iter := db.Iterator()
+//  for {
+//     key, value := iter.Key(), iter.Value()
+//     fmt.Printf("key: %v value: %v\n", key, value)
+//  }
+//
+// For better performance, perform Sets/Writes in batches
+//  batch := db.NewBatch()
+//  for i := 0; i < 100; i++ {
+//	batch.Set(cmn.RandBytes(100), []byte("true!!!"))
+//  }
+//  batch.Write()
+
+package db
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/dgraph-io/badger"
+
+	"github.com/tendermint/tmlibs/common"
+)
+
+func init() {
+	registerDBCreator(BadgerDBBackendStr, badgerDBCreator, true)
+}
+
+type Options badger.Options
+
+func badgerDBCreator(dbName, dir string) (DB, error) {
+	return NewBadgerDB(dbName, dir)
+}
+
+var (
+	_KB = int64(1024)
+	_MB = 1024 * _KB
+	_GB = 1024 * _MB
+)
+
+// NewBadgerDB creates a Badger key-value store backed to the
+// directory dir supplied. If dir does not exist, we create it.
+func NewBadgerDB(dbName, dir string) (*BadgerDB, error) {
+	// BadgerDB doesn't expose a way for us to
+	// create  a DB with the user's supplied name.
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return nil, err
+	}
+	opts := new(badger.Options)
```

Comment:
what if we just do `opts := badger.DefaultOptions`?

Reply:
For some reason, I initially thought `badger.DefaultOptions` was a pointer but somewhere I noticed not and left that in, lol. I'll update it, thank you!